### PR TITLE
ref(hc): Hybrid Cloud Authentication meets replication

### PR DESCRIPTION
**Pull request by corps**
Greatly reduces complexity of logic associated with hybrid cloud authentication, and begins the steps of normalizing `request.user` as an `RpcUser` in all cases.

1.  Request Framework Authentication (token auth) uses Replica models in Region mode.
a.  Has options to cut over normalizing monolith mode tokens into AuthenticatedTokens, so behavior between silos and saas is symmetrical.  This will carefully be rolled out, one auth method at a time.
b.  Includes options to cut over to using RpcUser in all cases (no more user objects in request).  Also will roll this out slowly.

Django Session Auth and Social Auth also swapped over to using RpcUser (behind option).

Removal of control level authentication logic (reduction of 470 lines!) thanks to being able to rely on replication.

### CODE REVIEW

File: src/sentry/models/integrations/organization_integration.py [hunk 1]
```diff
@@ -33,30 +38,25 @@ class OrganizationIntegration(DefaultFieldsModel):
...
+
+    @classmethod
+    def handle_async_deletion(
+        cls,
+        identifier: int,
+        region_name: str,
+        shard_identifier: int,
+        payload: Mapping[str, Any] | None,
+    ) -> None:
+        pass
```
Comment (markstory) [inline src/sentry/models/integrations/organization_integration.py:59]: :clap: Nice.

File: src/sentry/receivers/outbox/region.py [hunk 2]
```diff
@@ -41,7 +35,7 @@ def process_audit_log_event(payload: Any, **kwds: Any):
         log_rpc_service.record_audit_log(event=AuditLogEvent(**payload))
 
 
-@receiver(process_region_outbox, sender=OutboxCategory.ORGAUTHTOKEN_UPDATE)
+@receiver(process_region_outbox, sender=OutboxCategory.UPDATE_ORGAUTHTOKEN_USED)
```
Comment (markstory) [inline src/sentry/receivers/outbox/region.py:38]: Is `UPDATE_ORGAUTHTOKEN_USED` a permanent name? It doesn't follow the naming conventions that other outbox categories to of `thing_action`.

**Review by markstory — COMMENTED** [review]
There are a few different things going on in these changes, we might need to incrementally release this work to minimize risk.

File: src/sentry/receivers/outbox/region.py (hunk 2 above)
Comment (corps) [reply src/sentry/receivers/outbox/region.py:38]: Good call, I can rename easily.

**Comment by corps** [comment]
> There are a few different things going on in these changes, we might need to incrementally release this work to minimize risk.

I'd like to negotiate on this a bit more.

1.  I think the sizing is correct because it demonstrates the ability to fully remove the hybrid cloud authentication layer as it had existed.  Conditionally extracting specific authenticators is harder than it looks.

2.  The net effect is no change of behavior for SaaS.  The refactoring impact for the original saas behaviors are only the adjusted codepaths, but most of the new logic is additive and only activated for silo modes by default.

3.  I have hand validated existing test coverage for each of these changes.

I think the risk is lower on impact thanks to the sizable amounts of custom logic it replaces (logic which wasn't running in saas anyways, and logic that was more complicated than the new approaches).

File: src/sentry/api/authentication.py [hunk 3]
```diff
@@ -287,37 +340,59 @@ def accepts_auth(self, auth: list[bytes]) -> bool:
...
-                .select_related("user", "application")
-                .get()
-            )
-        except ApiToken.DoesNotExist:
+        user: AnonymousUser | RpcUser | None = AnonymousUser()
+
+        token: SystemToken | ApiTokenReplica | ApiToken | None = SystemToken.from_request(
+            request, token_str
+        )
+        application_is_active = True
```
Comment (corps) [inline src/sentry/api/authentication.py:348]: `application_is_active = False` here, then `application_is_active` on line 369

File: src/sentry/api/authentication.py [hunk 4]
```diff
@@ -287,37 +340,59 @@ def accepts_auth(self, auth: list[bytes]) -> bool:
...
+                        .get()
+                    )
+                except ApiToken.DoesNotExist:
+                    raise AuthenticationFailed("Invalid token")
+                user = at.user
+                application_is_active = at.application is None or at.application.is_active
+        elif isinstance(token, SystemToken):
+            user = token.user
+
+        if not token:
```
Comment (corps) [inline src/sentry/api/authentication.py:371]: `else`

**Review by asottile-sentry — COMMENTED** [review]
learnings from our past outages:
- this patch is too big to be reviewable
- migrations should be split into separate PRs

File: src/sentry/api/authentication.py [hunk 5]
```diff
@@ -292,37 +341,59 @@ def accepts_auth(self, auth: list[bytes]) -> bool:
...
+                try:
+                    at = token = (
+                        ApiToken.objects.filter(token=token_str)
+                        .select_related("user", "application")
+                        .get()
+                    )
+                except ApiToken.DoesNotExist:
+                    raise AuthenticationFailed("Invalid token")
+                user = at.user
+                application_is_active = at.application is None or at.application.is_active
```
Comment (markstory) [inline src/sentry/api/authentication.py:368]: If the token has no `.application` shouldn't `application_is_active` be False?  I can see how the end result is correct based on the check on 381, but this being true when the application doesn't exist feels like a footgun for future us.

File: src/sentry/api/authentication.py [hunk 6]
```diff
@@ -292,37 +341,59 @@ def accepts_auth(self, auth: list[bytes]) -> bool:
...
 
         if token.is_expired():
             raise AuthenticationFailed("Token expired")
 
-        if not token.user.is_active:
+        if user and hasattr(user, "is_active") and not user.is_active:
             raise AuthenticationFailed("User inactive or deleted")
 
-        if token.application and not token.application.is_active:
+        if not application_is_active:
```
Comment (markstory) [inline src/sentry/api/authentication.py:381]: Perhaps we could extract `token.application` into a local var, and then be able to do `if application and not application.is_active` to avoid the incorrect local var.

File: src/sentry/api/authentication.py [hunk 7]
```diff
@@ -127,9 +132,59 @@ def relay_from_id(request, relay_id) -> Tuple[Optional[Relay], bool]:
...
+        auth_token = AuthenticatedToken.from_token(auth)
+        if auth_token and entity_id_tag:
+            with configure_scope() as scope:
+                scope.set_tag(entity_id_tag, auth_token.entity_id)
+                for k, v in tags.items():
+                    scope.set_tag(k, v)
+
+        return (
+            user,
+            AuthenticatedToken.from_token(auth)
```
Comment (markstory) [inline src/sentry/api/authentication.py:183]: Are we able to reuse `auth_token` given that we already made one?

File: src/sentry/api/authentication.py [hunk 8]
```diff
@@ -127,9 +132,59 @@ def relay_from_id(request, relay_id) -> Tuple[Optional[Relay], bool]:
...
+            SiloMode.get_current_mode() != SiloMode.MONOLITH
+            or options.get("hybrid_cloud.authentication.use_rpc_user")
+            >= self._hybrid_cloud_rollout_level
+            or in_test_environment()
+        )
+
+    def transform_auth(
+        self,
+        user: int | User | RpcUser | None | AnonymousUser,
+        auth: Any,
```
Comment (markstory) [inline src/sentry/api/authentication.py:159]: We overload `auth` often, could we use a name like `auth_token` 🤷

File: src/sentry/deletions/defaults/apigrant.py [hunk 9]
```diff
@@ -1,3 +1,7 @@
+from django.db import router
+
+from ...models import ApiGrant
+from ...silo import unguarded_write
```
Comment (markstory) [inline src/sentry/deletions/defaults/apigrant.py:4]: Should these imports be absolute?

File: src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+                        "sentry.ApiApplication", db_index=True, null=True, on_delete="CASCADE"
+                    ),
+                ),
+                (
+                    "organization_id",
+                    sentry.db.models.fields.hybrid_cloud_foreign_key.HybridCloudForeignKey(
+                        "sentry.Organization", db_index=True, null=True, on_delete="SET_NULL"
+                    ),
+                ),
+                ("application_is_active", models.BooleanField(default=False)),
```
Comment (markstory) [inline src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py:80]: Does this field need to be kept? I didn't see it in use by the authentication adapters.

File: src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+                    "application_id",
+                    sentry.db.models.fields.hybrid_cloud_foreign_key.HybridCloudForeignKey(
+                        "sentry.ApiApplication", db_index=True, null=True, on_delete="CASCADE"
+                    ),
+                ),
+                (
+                    "organization_id",
+                    sentry.db.models.fields.hybrid_cloud_foreign_key.HybridCloudForeignKey(
+                        "sentry.Organization", db_index=True, null=True, on_delete="SET_NULL"
+                    ),
```
Comment (markstory) [inline src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py:78]: If this is a region bound model what does the HybridCloudForeignKey here?

File: src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+                ),
+                ("application_is_active", models.BooleanField(default=False)),
+                (
+                    "user_id",
+                    sentry.db.models.fields.hybrid_cloud_foreign_key.HybridCloudForeignKey(
+                        "sentry.User", db_index=True, on_delete="CASCADE"
+                    ),
```
Comment (markstory) [inline src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py:85]: I don't think user_id can be unique. Wouldn't that limit users to only having a single apitoken?

File: src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+                ),
+                (
+                    "apitoken_id",
+                    sentry.db.models.fields.hybrid_cloud_foreign_key.HybridCloudForeignKey(
+                        "sentry.ApiToken", db_index=True, on_delete="CASCADE"
+                    ),
+                ),
+                ("token", models.CharField(max_length=64)),
```
Comment (markstory) [inline src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py:93]: Should this column be unique so that we can't have overlapping tokens even if replication goes sideways?

File: src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+                        primary_key=True, serialize=False
+                    ),
+                ),
+                (
+                    "orgauthtoken_id",
+                    sentry.db.models.fields.hybrid_cloud_foreign_key.HybridCloudForeignKey(
+                        "sentry.OrgAuthToken", db_index=True, on_delete="CASCADE"
+                    ),
+                ),
+                ("token_hashed", models.TextField()),
```
Comment (markstory) [inline src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py:117]: Should this be unique=True to prevent duplicates even if replication goes sideways?

File: src/sentry/web/frontend/setup_wizard.py [hunk 15]
```diff
@@ -117,7 +117,7 @@ def get_token(orgs: list[Organization], user: User):
             return token
 
     # Otherwise, generate a user token
-    tokens = ApiToken.objects.filter(user=user)
+    tokens = ApiToken.objects.filter(user_id=user.id)
```
Comment (markstory) [inline src/sentry/web/frontend/setup_wizard.py:120]: Having to use `user_id` instead of `user` as a filter when you're working with the request.user is something we should tell folks more broadly about.

File: src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py (hunk 12 above)
Comment (corps) [reply src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py:85]: It's not unique, it just has an index right?

File: src/sentry/web/frontend/setup_wizard.py (hunk 15 above)
Comment (corps) [reply src/sentry/web/frontend/setup_wizard.py:120]: Yup, that will be part of this rollout for sure.

Comment (corps) [reply src/sentry/web/frontend/setup_wizard.py:120]: Notably, after this PR lands, tests will reflect RpcUser and these issues will become more obvious immediately, even if the rollout in production is staged.

File: src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py (hunk 11 above)
Comment (corps) [reply src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py:78]: Good point, this can be a flexible foreign key instead, will adjust.

File: src/sentry/api/authentication.py (hunk 6 above)
Comment (corps) [reply src/sentry/api/authentication.py:381]: probably yes, will refactor.

Comment (corps) [reply src/sentry/api/authentication.py:381]: took another pass on this -- I inverted the condition to make it easier.  The reason why extracting application and using is_active doesn't work great is that ApiTokenReplica has the is_application_active attribute and not a reference to the application directly.  This is because in region mode an application reference is only obtainable either via RPC or replicating that entire object (which seemed overkill).

File: src/sentry/api/authentication.py (hunk 7 above)
Comment (corps) [reply src/sentry/api/authentication.py:183]: Fixed

File: src/sentry/api/authentication.py (hunk 8 above)
Comment (corps) [reply src/sentry/api/authentication.py:159]: Yes -- but, `auth` is the canonical word for the the thing we're expecting here (whatever went into `request.auth`.  What if I call it `request_auth`?

File: src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py (hunk 13 above)
Comment (corps) [reply src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py:93]: Not quite.  I've taken the strategy of keeping this non unique so that replication, in the case of a conflict, succeeds eventually.  On the query side, I take the `.last()` item.  This greatly reduces conflict risks.

Comment (markstory) [reply src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py:93]: Isn't the risk of duplicate tokens very high though? While it shouldn't happen in practice having duplicate tokens and choosing one of them could result in customers gaining access to organizations they shouldn't.

File: src/sentry/api/authentication.py (hunk 8 above)
Comment (markstory) [reply src/sentry/api/authentication.py]: request_auth works for me.

File: src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py (hunk 13 above)
Comment (corps) [reply src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py:93]: @markstory To be clear, it's worth recalling the exact scenario where that would happen: in the world in which on control, at some point, *that token did have access*.  This will always be a problem technically -- in the face of partition, a token in the past that was valid but is not in the present valid, may still be presented.  The replication doesn't change the security story here.  Actually, by leaving this index non unique, it reduces any chance of locking up replication (because the column isn't along the shard column we can't use the delete trick we do for other unique columns), so I believe that strictly speaking it is better to take this approach.

Basically, it's worth remembering that not enforcing uniqueness is not the same as letting arbitrary tokens leak.  It's really ultimately how we handle the hand off of two valid states in different points in time, and no matter what implementation we use, eventual consistency will be an issue.

Comment (corps) [reply src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py:93]: Notably too: I can guarantee that so long as we don't treat this as a unique column, we always do converge on uniqueness with respect to the original source, it's only the potential of a hand off (circular) state that presents any complexity.

Comment (markstory) [reply src/sentry/hybridcloud/migrations/0005_add_api_and_org_token_replicas.py:93]: :+1: Thanks for the reminder about potentially locking up outboxes. Relaxing the constraint is our best option to avoid that, and we need to rely on the origin table's constraint to ensure that we'll eventually settle into a state that doesn't have duplicate tokens.

File: src/sentry/middleware/access_log.py [hunk 16]
```diff
@@ -11,6 +11,7 @@
 
 from sentry.utils import metrics
 
+from ..services.hybrid_cloud.auth import AuthenticatedToken
```
Comment (markstory) [inline src/sentry/middleware/access_log.py:14]:
```suggestion
from sentry.services.hybrid_cloud.auth import AuthenticatedToken
```

File: src/sentry/models/outbox.py [hunk 17]
```diff
@@ -103,6 +103,8 @@ class OutboxCategory(IntEnum):
     ORGANIZATION_SLUG_RESERVATION_UPDATE = 27
     API_KEY_UPDATE = 28
     PARTNER_ACCOUNT_UPDATE = 29
+    API_TOKEN_UPDATE = 30
+    ORG_AUTH_TOKEN_UPDATE = 31
```
Comment (markstory) [inline src/sentry/models/outbox.py:107]: :+1: to a new category to avoid crossed paths during replication.

**Review by markstory — COMMENTED** [review]
I think this looks good. Lets get the migration shipped.

File: tests/sentry/middleware/test_access_log_middleware.py [hunk 18]
```diff
@@ -183,7 +183,7 @@ def test_access_log_success(self):
         self.login_as(user=self.create_user())
         self.get_success_response(extra_headers={"HTTP_AUTHORIZATION": f"Bearer {token.token}"})
         self.assert_access_log_recorded()
-        assert self.captured_logs[0].token_type == "ApiToken"
+        assert self.captured_logs[0].token_type == "api_token"
```
Comment (mdtro) [inline tests/sentry/middleware/test_access_log_middleware.py:186]: @Milkshak3s Just a heads up. It looks like the value for `token_type` in the logs query/use will be changing from `ApiToken` to `api_token`.

Comment (corps) [reply tests/sentry/middleware/test_access_log_middleware.py:186]: @mdtro Yes -- as per this PR and moving forward, the original token class(es) get replaced with a single unified model `AuthenticatedToken` which maps back to this casing of the source of that token.  Is the casing important?  I could convert all of my output types to retain compatibility but ideally, the casing helps clarify this difference between storage types (`ApiToken`) and representation types (`AuthenticatedToken.token_type == 'api_token'`).  This is important because the backing storage can also differ based on silo (`ApiTokenReplica` for region silos)

File: src/sentry/api/authentication.py [hunk 19]
```diff
@@ -191,42 +246,30 @@ def authenticate_credentials(self, relay_id, relay_sig, request):
...
-            return AnonymousUser(), self._authenticate_credentials(userid)
-
-        try:
-            key = ApiKey.objects.get_from_cache(key=userid)
-        except ApiKey.DoesNotExist:
-            raise AuthenticationFailed("API key is not valid")
+        if SiloMode.get_current_mode() == SiloMode.REGION:
+            key = ApiKeyReplica.objects.filter(key=userid).last()
+            if key is None:
+                raise AuthenticationFailed("API key is not valid")
```
Comment (GabeVillalobos) [inline src/sentry/api/authentication.py:262]: Would there be any value in moving this to conditional model lookup to the auth service impl instead?

File: src/sentry/api/authentication.py [hunk 20]
```diff
@@ -292,37 +341,62 @@ def accepts_auth(self, auth: list[bytes]) -> bool:
...
+                        ApiToken.objects.filter(token=token_str)
+                        .select_related("user", "application")
+                        .get()
+                    )
+                except ApiToken.DoesNotExist:
+                    raise AuthenticationFailed("Invalid token")
+                user = at.user
+                application_is_inactive = (
+                    at.application is not None and not at.application.is_active
+                )
```
Comment (GabeVillalobos) [inline src/sentry/api/authentication.py:371]: Same comment on this being possibly moved to the auth service impl

Comment (corps) [reply src/sentry/api/authentication.py:371]: Hmmm.... I guess, not?  My intention here is to keep the full implementation consolidated.  The "auth service" really has much to do with the RPC elements of the HC implementation.  My hope here is to ensure that the singular Token lookup logic is in one, clear place, and not divided.  For instance, it really isn't HC's job to say anything about ho tokens are used for auth, and that's the goal of this PR.  That said, I'm open to refactoring these conditionals into separate methods if that would look better.

File: src/sentry/services/hybrid_cloud/auth/model.py [hunk 21]
```diff
@@ -215,27 +95,28 @@ def from_token(cls, token: Any) -> Optional["AuthenticatedToken"]:
...
             if any(isinstance(token, kind_cls) for kind_cls in types):
                 break
         else:
             raise KeyError(f"Token {token} is a not a registered AuthenticatedToken type!")
 
+        entity_id: Optional[int] = None
+        # System tokens have a string id but don't really represent a true entity
+        if kind != "system":
+            entity_id = getattr(token, "entity_id", getattr(token, "id", None))
+
```
[the remark this answers is not in the document]
Comment (corps) [reply src/sentry/services/hybrid_cloud/auth/model.py:108]: Added tests

**Review by GabeVillalobos — APPROVED** [review]
